### PR TITLE
Bozin Dmitry Lab3 Var1

### PR DESCRIPTION
**Pull request by Nilemos**

### CODE REVIEW

File: llvm/lib/Target/X86/lab3/bozin_dmitry/X86BozinPass.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+              reg_addic_1 = addic_instr->getOperand(1).getReg();
+              reg_addic_2 = addic_instr->getOperand(2).getReg();
+              if (reg_multi == reg_addic_1 || reg_multi == reg_addic_2) {
+                delete_instr.emplace_back(multi_instr, addic_instr);
+                parameter = true;
+                if (reg_multi == reg_addic_1) {
+                  reg = false;
+                } else {
+                  reg = true;
+                }
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/bozin_dmitry/X86BozinPass.cpp:45]:
```suggestion
                reg = (reg_multi != reg_addic_1);
```

File: llvm/lib/Target/X86/lab3/bozin_dmitry/X86BozinPass.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    for (auto &MBB : MF) {
+      MachineInstr *multi_instr = nullptr;
+      MachineInstr *addic_instr = nullptr;
+      Register reg_multi;
+      Register reg_addic_1;
+      Register reg_addic_2;
+      for (auto &instr : MBB) {
+        if (instr.getOpcode() == X86::MULPDrr) {
+          multi_instr = &instr;
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/bozin_dmitry/X86BozinPass.cpp:30]:
```suggestion
        if (instr.getOpcode() != X86::MULPDrr) {
          continue;
        multi_instr = &instr;
```

Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/bozin_dmitry/X86BozinPass.cpp:30]:
```suggestion
        if (instr.getOpcode() != X86::MULPDrr) {
          continue;
        multi_instr = &instr;
```

File: llvm/test/lab3/bozin_dmitry/tests.mir [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+constants:       []
+machineFunctionInfo: {}
+body:             |
+  bb.0.entry:
+    liveins: $xmm0, $xmm1
+  
+    renamable $xmm0 = nofpexcept MULPDrr killed renamable $xmm0, killed renamable $xmm1, implicit $mxcsr
+    renamable $xmm0 = nofpexcept ADDPDrr killed renamable $xmm0, renamable $xmm0, implicit $mxcsr
+    RET64 $xmm0
+    ; CHECK: $xmm0 = VFMADD213PDZ128r $xmm0, $xmm1, $xmm0, implicit $mxcsr
```
Comment (m-ly4) [inline llvm/test/lab3/bozin_dmitry/tests.mir:493]: This is equal to `a * b + a` whilst initial expression was `a * b + a * b`.

File: llvm/test/lab3/bozin_dmitry/tests.mir [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+#   __m128d tmp = a * b;
+#   __m128d tmp2 = b + a - c;
+#   return tmp + c + tmp2;
+# }
+
+# __m128d multaddc_test6(__m128d a, __m128d b, __m128d c) {
+#   __m128d tmp = a * b;
+#   return tmp + tmp;
+# }
+
```
Comment (m-ly4) [inline llvm/test/lab3/bozin_dmitry/tests.mir:35]: Please cover also the following case:
```
# __m128d multaddc_test4(__m128d a, __m128d b, __m128d c) {
#   __m128d tmp = a * c;
#   __m128d tmp2 = tmp + b;
#   return tmp / b;
# }
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/X86/lab3/bozin_dmitry/X86BozinPass.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace {
+class X86MulAddPass : public MachineFunctionPass {
+public:
+  static char ID;
+  X86MulAddPass() : MachineFunctionPass(ID) {}
+  bool runOnMachineFunction(MachineFunction &MF) override {
+    const TargetInstrInfo *info = MF.getSubtarget().getInstrInfo();
+    std::vector<std::pair<MachineInstr *, MachineInstr *>> delete_instr;
+    bool parameter = false;
```
Comment (aobolensk) [inline llvm/lib/Target/X86/lab3/bozin_dmitry/X86BozinPass.cpp:19]: What is "parameter"?

File: llvm/lib/Target/X86/lab3/bozin_dmitry/X86BozinPass.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                  info->get(X86::VFMADD213PDZ128r));
+      builder.addReg(addc->getOperand(0).getReg(), RegState::Define);
+      builder.addReg(mult->getOperand(1).getReg());
+      builder.addReg(mult->getOperand(2).getReg());
+      if (reg) {
+        builder.addReg(addc->getOperand(1).getReg());
+      } else {
+        builder.addReg(addc->getOperand(2).getReg());
+      }
+      mult->eraseFromParent();
```
Comment (aobolensk) [inline llvm/lib/Target/X86/lab3/bozin_dmitry/X86BozinPass.cpp:69]: Don't we need to check for "mult" uses before removal?

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Pull request closed without merging**